### PR TITLE
732 notif collab resolve

### PR DESCRIPTION
**Pull request by drewblount**
Collaborations now update the collab request notifications with `is_resolved=True` when those requests are accepted/denied. It keeps track of a couple notification guids in the collab model to facilitate that. The updating logic is in Collaboration._notify_users. A new test in test_collaboration_notification checks that this gets updated.

Collaboration.__init__ has been updated to automatically send the request notification via Collaboration.notify_pending_users(), which previously had to be called manually by the code wherever the collab was made. This auto-send can be disabled by passing in `notify_users=False`. Doing this automatically during collab init is a good idea because there were a few tests / places in code where we had mixed up who is the sender vs. recipient of both the Collab and the Notif; putting both in one call means we can't mix them up.

### CODE REVIEW

File: app/modules/collaborations/models.py [hunk 1]
```diff
@@ -92,11 +92,17 @@ class Collaboration(db.Model, HoustonModel):
     initiator_guid = db.Column(
         db.GUID, db.ForeignKey('user.guid'), index=True, nullable=False
     )
+    init_req_notification_guuid = db.Column(
```
Comment (hwindsor) [inline app/modules/collaborations/models.py:95]: guid not guuid, that will confuse

File: app/modules/collaborations/models.py [hunk 2]
```diff
@@ -92,11 +92,17 @@ class Collaboration(db.Model, HoustonModel):
...
+        db.GUID, db.ForeignKey('notification.guid'), nullable=True
+    )
     edit_initiator_guid = db.Column(
         db.GUID, db.ForeignKey('user.guid'), index=True, nullable=True
     )
+    edit_req_notification_guuid = db.Column(
```
Comment (hwindsor) [inline app/modules/collaborations/models.py:101]: likewise

File: app/modules/collaborations/models.py [hunk 3]
```diff
@@ -221,11 +232,47 @@ def notify_pending_users(self):
...
+            Notification,
+            NotificationBuilder,
+            NotificationType,
+        )
 
         builder = NotificationBuilder(sending_user_assoc.user)
         builder.set_collaboration(self)
-        Notification.create(notification_type, receiving_user_assoc.user, builder)
+        notif = Notification.create(notification_type, receiving_user_assoc.user, builder)
+        # I expect this to be set by the builder but it sometimes does not, unsure why.
```
Comment (hwindsor) [inline app/modules/collaborations/models.py:244]: It really should have been. Curious about when you were seeing it unset, was it when running uit tests?

File: app/modules/collaborations/models.py [hunk 4]
```diff
@@ -221,11 +232,47 @@ def notify_pending_users(self):
...
+        elif notification_type is NotificationType.collab_approved:
+            if self.init_req_notification_guuid:
+                self._resolve_notification(self.init_req_notification_guuid)
+        elif notification_type in fully_resolved_notification_states:
+            if self.init_req_notification_guuid:
+                self._resolve_notification(self.init_req_notification_guuid)
+            if self.edit_req_notification_guuid:
+                self._resolve_notification(self.edit_req_notification_guuid)
+
+    def _resolve_notification(self, notification_guid):
```
Comment (hwindsor) [inline app/modules/collaborations/models.py:269]: As a guiding principle, I think we should have modifications of an object done within that object, so this should be a classmethod in the Notification class.

**Review by hwindsor — APPROVED** [review]
Generally all good, 
Couple of minor tweaks requested.
Thanks for sorting out those tests, I was obviously having a brain fade when I wrote them.

File: app/modules/collaborations/models.py (hunk 3 above)
Comment (drewblount) [reply app/modules/collaborations/models.py:244]: Glad you double checked, I thought one of the tests broke without that line but was not able to replicate just now. Must have been one of those debugging red herrings, so I'm removing this line.

**Review by hwindsor — APPROVED** [review]
Thanks for that, much better, just need to merge those migrations (downgrade, delete both and rerun the migration)

**Comment by drewblount** [comment]
\rebase

**Pull request merged**